# Modeling the Effects of Gene Perturbations inÂ Maize

<p align="center">
  Contributors: Victor Avram, Sergio Jimenez, Eagon Meng, Wenhan Zhang
</p>
<p align="center">
  <img src="https://user-images.githubusercontent.com/29682604/117849295-2bfd1200-b252-11eb-8d74-c186ab51c77e.png">
</p>

## Introduction and Motivation
---

As a plant breeding technology company, Inari's work comprises three high-level stages: (1) using computational methods to better understand biology, especially at the transcriptional level; (2) genetic editing of crops; and (3) delivery of altered genetic information to specific parts of the plant. By genetically modifying seeds, Inari seeks to increase crop diversity, increase yield so as to make efficient use of land, water, and fertilizer, and ultimately make socio-economic impacts by ensuring food security.

Our project sits within the first stage of Inari's work and seeks to answer "what are the effects of gene perturbations and are gene expression levels informative of one another?" Tackling these questions would provide Inari with a better understanding maize biology and would subsequently allow seed genetic editing with greater confidence. Essentially, we hope to document the relationships across genes, creating a network that can act as a look-up table to inform Inari of the genetic effects and side-effects of perturbing particular maize genes. These insights will help alleviate the risk of observing unexpected effects later in the breeding process. We have explored several methods to determine whether a subset (ideally a small subset) of genes can be used as predictors for the expression levels of the remaining genes in the maize genome. Through this process, we are able to realize relationships across genes and ultimately allow for more targeted gene modifications.

## Data and EDA
---

## Feature Selection and Regression-Based Methods
---

### Standard Feature Selection + Regression

### Landmark 1000 + Regression

## Autoencoder-Based Methods
---

### Autoencoder + Regression: Learning Structures of Expression Data

### Imputation using an Iterative Variational Autoencoder

## Graph Neural Networks
---

## References

